spring-integration-core/src/main/java/org/springframework/integration/handler/MessageHandlerChain.java

```diff
@@ -64,7 +64,7 @@
  * @author Artem Bilan
  */
 public class MessageHandlerChain extends AbstractMessageProducingHandler implements MessageProducer,
-		CompositeMessageHandler, Lifecycle {
+		HeaderPropagationAware, CompositeMessageHandler, Lifecycle {
```

Comment:
That's not true.
We don't propagate headers only in this case:
```
if (this.shouldCopyRequestHeaders()) {
		if (this.selectiveHeaderPropagation) {
			Map<String, Object> headersToCopy = new HashMap<String, Object>(requestHeaders);
			for (String header : this.notPropagatedHeaders) {
				headersToCopy.remove(header);
			}
			builder.copyHeadersIfAbsent(headersToCopy);
		}
```
So, or we shouldn't care about `MessageHandlerChain`  for this task (including any other should makes `shouldCopyRequestHeaders()` as `false`), or we should have some more sophisticated logic and clean the reply message from those headers independently of the component type. Pay attention to the `BridgeHandler` BTW...